[mlir][ArmSME] Support 2-way widening outer products

c-rhodes opened the pull request:
This patch introduces support for 2-way widening outer products. This
enables the fusion of 2 'arm_sme.outerproduct' operations that are
chained via the accumulator into a 2-way widening outer product operation.

Changes:

- Add 'llvm.aarch64.sme.[us]mop[as].za32' intrinsics for 2-way variants.
  These map to instruction variants added in SME2 and use different
  intrinsics. Intrinsics are already implemented for widening variants
  from SME1.
- Adds the following operations:
  - fmopa_2way, fmops_2way
  - smopa_2way, smops_2way
  - umopa_2way, umops_2way
- Implements conversions for the above ops to intrinsics in ArmSMEToLLVM.
- Adds a pass 'arm-sme-outer-product-fusion'  that fuses
  'arm_sme.outerproduct' operations.

For a detailed description of these operations see the
'arm_sme.fmopa_2way' description.

The reason for introducing many operations rather than one is the
signed/unsigned variants can't be distinguished with types (e.g., ui16,
si16) since 'arith.extui' and 'arith.extsi' only support signless
integers. A single operation would require this information and an
attribute (for example) for the sign doesn't feel right if
floating-point types are also supported where this wouldn't apply.
Furthermore, the SME FP8 extensions (FEAT_SME_F8F16, FEAT_SME_F8F32)
introduce FMOPA 2-way (FP8 to FP16) and 4-way (FP8 to FP32) variants but
no subtract variant. Whilst these are not supported in this patch, it
felt simpler to have separate ops for add/subtract given this.

c-rhodes committed:
[mlir][ArmSME] Support 2-way widening outer products

This patch introduces support for 2-way widening outer products. This
enables the folding of 2 'arm_sme.outerproduct' operations that are
chained via the accumulator into single widened operations.

Changes:

- Add 'llvm.aarch64.sme.[us]mop[as].za32' intrinsics for 2-way variants.
  These map to instruction variants added in SME2 and use different
  intrinsics. Intrinsics are already implemented for widening variants
  from SME1.
- Adds the following operations:
  - fmopa_wide_2way, fmops_wide_2way
  - smopa_wide_2way, smops_wide_2way
  - umopa_wide_2way, umops_wide_2way
- Implements conversions for the above ops to intrinsics in ArmSMEToLLVM.
- Adds a pass 'arm-sme-outer-product' widening that folds
  'arm_sme.outerproduct' operations.

For a detailed description of these operations see the
'arm_sme.fmopa_wide_2way' description.

The reason for introducing many operations rather than one is the
signed/unsigned variants can't be distinguished with types (e.g., ui16,
si16) since 'arith.extui' and 'arith.extsi' only support signless
integers. A single operation would require this information and an
attribute (for example) for the sign doesn't feel right if
floating-point types are also supported where this wouldn't apply.
Furthermore, the SME FP8 extensions (FEAT_SME_F8F16, FEAT_SME_F8F32)
introduce FMOPA 2-way (FP8 to FP16) and 4-way (FP8 to FP32) variants but
no subtract variant. Whilst these are not supported in this patch, it
felt simpler to have separate ops for add/subtract given this.

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -736,6 +736,7 @@ class OuterProductResultTileTypeConstraint<string operand> :
 
 def OuterProductOp :
   ArmSME_Op<"outerproduct", [
+    Pure,
```
No ArmSME tile ops are pure, as they're implicitly working on ZA, which is not modeled. 

Consider: 
```
%0 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1 (allocs a new tile)
%1 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1 (allocs a new tile)
// Move tile slice into %1 (assume slice is all 0)
%2 = arm_sme.move_vector_to_tile_slice %1 ....
arm_sme.tile_store %0 ...
arm_sme.tile_store %2 ...
```
This results in a store of `%0` which is all 1, followed by a store of `%2`, which has one row of zeros.

Now since `arm_sme.outerproduct` is pure, cse can rewrite this to:
```
%0 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1
// Move tile slice into %1 (assume slice is all 0)
%1 = arm_sme.move_vector_to_tile_slice %0 ....
arm_sme.tile_store %0 ...
arm_sme.tile_store %1 ...
```
In this case all operations (after tile allocation) are working on the same tile. So _both_ will result in a store of a tile with one slice of zeros, which is incorrect.  

I don't think this can be fixed without proper tile value semantics.

c-rhodes replied:
ah, good catch, I'm thinking of this op like a Vector dialect op and of course that's not the case. I'll remove this, it's not a problem because in the pass I'm explicitly checking all but the last outer product in a chain have single use and erasing them. Marking this as pure means cse would do that for us, but it's not necessary.

MacDue commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -814,6 +815,650 @@ let arguments = (ins
+                           list<Type> allowedResultVectorTypes,
+                           int numOuterProducts> :
+  ArmSME_Op<mnemonic, [
+    Pure,
```
See above.

MacDue reviewed (commented):
Just a quick comment, will have a proper look later:

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -814,6 +814,649 @@ let arguments = (ins
+    | ---- | -------- |
+    | [FMOPA (widening, 2-way, FP16 to FP32)](https://developer.arm.com/documentation/ddi0602/2023-09/SME-Instructions/FMOPA--widening--2-way--FP16-to-FP32---Half-precision-floating-point-sum-of-outer-products-and-accumulate-) | +sme |
+    | [BFMOPA (widening, 2-way, BF16 to FP32)](https://developer.arm.com/documentation/ddi0602/2023-09/SME-Instructions/BFMOPA--widening---BFloat16-sum-of-outer-products-and-accumulate-) | +sme |
+me
```
Is this meant to be here?

banach-space replied:
I meant specifically `me` on L993 :)

c-rhodes replied:
argh, no it's not :)

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -814,6 +814,649 @@ let arguments = (ins
+  let summary = "Floating-point sum of 2 outer products and accumulate";
+
+  let description = [{
+    This operation represents a sum of 2 widened outer products. It takes 2 1-D
+    scalable vectors as input and a 2-D scalable vector (ZA tile) as output.
```
```suggestion
    This operation represents a sum of 2 widening outer products. It takes 2 x 1D
    scalable vectors as input and a 2D scalable vector (ZA tile) as output.
```

c-rhodes replied:
`1-D` / `2-D` are more common so I'll keep that format. Also not sure about `widened` vs `widening` but instructions descriptions in the ISA use widened so I'll just keep it as that.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
````diff
@@ -814,6 +814,649 @@ let arguments = (ins
+
+    ```
+                          LHS                          RHS
+               [A0 A1 A2 A3 A4 A5 A6 A7]    [B0 B1 B2 B3 B4 B5 B6 B7]
````
How "wide" are elements of LHS and RHS? (e.g. A0 and B0)?

banach-space replied:
I guess that's implicit in this description. I'd considering spelling that out. No drama if not.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
````diff
@@ -814,6 +814,649 @@ let arguments = (ins
+
+	```mlir
+    %undef = llvm.mlir.undef : vector<[8]xf16>
+    %a0_ins = vector.scalable.ins %a0_ext, %undef[0] : vector<[4]xf16> into vector<[8]xf16>
````
```suggestion
    %a0_ins = vector.scalable.insert %a0_ext, %undef[0] : vector<[4]xf16> into vector<[8]xf16>
```
?

MacDue reviewed (commented):
Few more comments, will look again soon, not looked at everything yet :)

Also, not sure if you've noticed but this test is failing in CI:
```
Failed Tests (1):
  MLIR :: Dialect/ArmSME/outer-product-widening.mlir
```
(Looks like it's crashing during a rewrite)

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,501 @@
+            kind == arm_sme::CombiningKind::Sub) &&
+           "unhandled arm_sme::CombiningKind!");
+
+    if (isa<arith::ExtFOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::FMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else if (isa<arith::ExtSIOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::SMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else if (isa<arith::ExtUIOp>(extOp)) {
+      if (kind == arm_sme::CombiningKind::Add)
+        rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+      else
+        rewriter.replaceOpWithNewOp<arm_sme::UMopsWide2WayOp>(
+            op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
+    } else
+      llvm_unreachable("unexpected extend op!");
+
+    op1.erase();
+
+    return success();
```
IMHO, `else` obscures the flow. This might be a matter of style, so feel free to ignore.

```suggestion
if (kind == arm_sme::CombiningKind::Add) {
  if (isa<arith::ExtFOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtSIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtUIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  llvm_unreachable("unexpected extend op!");
}

if (kind == arm_sme::CombiningKind::Sub) {
  if (isa<arith::ExtFOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::FMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtSIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::SMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  if (isa<arith::ExtUIOp>(extOp)) {
    rewriter.replaceOpWithNewOp<arm_sme::UMopsWide2WayOp>(
        op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask, op1.getAcc());
  }
  llvm_unreachable("unexpected extend op!");
}

op1.erase();
return success();
```

MacDue replied:
Or an approach I like with TypeSwitch:
```c++
  if (kind == arm_sme::CombiningKind::Add) {
    TypeSwitch<Operation *>(extOp)
        .Case<cf::ExtFOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::FMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Case<cf::ExtSIOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::SMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Case<ExtUIOp>([&](auto) {
          rewriter.replaceOpWithNewOp<arm_sme::UMopaWide2WayOp>(
              op2, op.getResultType(), lhs, rhs, lhsMask, rhsMask,
              op1.getAcc());
        })
        .Default([&](auto) { llvm_unreachable("unexpected extend op!"); });
  }

  /// ... same again for Sub ...
```

MacDue commented on mlir/include/mlir/Dialect/ArmSVE/IR/ArmSVE.td:
```diff
@@ -410,4 +410,8 @@ def ConvertToSvboolIntrOp :
 
+def Zip1IntrOp :
+  ArmSVE_IntrBinaryOverloadedOp<"zip1">,
+  Arguments<(ins AnyScalableVector, AnyScalableVector)>;
```
> Why? As in, what's the difference between this and other intrinsics? Is it something from the ISA that I am missing?

There's no difference, I think they all should be using SVEVector as their types. I think these intrinsics (that very closely map to hardware instructions), have to have types that match actual hardware sizes, or they'll fail to lower: https://godbolt.org/z/o5c941Gb3 (as a random example).

c-rhodes replied:
You're right Ben these constraints don't match the ISA, although I should point out the intrinsic definitions in the backend are typically very relaxed as well, for zip1 see:

https://github.com/llvm/llvm-project/blob/b0763a1ae940d60d8f558f85216382bc6695a1e3/llvm/include/llvm/IR/IntrinsicsAArch64.td#L130-L132

https://github.com/llvm/llvm-project/blob/b0763a1ae940d60d8f558f85216382bc6695a1e3/llvm/include/llvm/IR/IntrinsicsAArch64.td#L1927

I suspect the intrinsic definitions in the SVE dialect have followed suit, that's not to say we can't do better in MLIR however, and for SME that's been the case, the MLIR constraints are tighter than the LLVM ones.

> Also, @c-rhodes , thinking how to split this massive patch - would you mind moving zip1 to a dedicated PR?

sure 👍

c-rhodes replied:
moved to #79270

c-rhodes commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,501 @@
+
+// Fold four 'arm_sme.outerproduct' operations that are chained via the
+// accumulator into 4-way outer product operation.
+class OuterProduct4WayWidening
```
sure 👍

c-rhodes replied:
moved to #79288

banach-space commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -814,6 +814,649 @@ let arguments = (ins
+                         allowedResultVectorTypes, /*numOuterProducts=*/4>;
+
+def FMopaWide2WayOp
+  : OuterProductWide2Way<"fmopa_wide_2way",
```
Yes, bit of a mouthful :)

I assume that all 2way and 4way MOPAs are "widening". If there's an exception, we could ad d `fmopa_nw_2way`at some later time.

MacDue replied:
Just saying I oppose cryptic names like `nw` (we really should not worry about the length operations names, which outside of unit tests are not human written).

MacDue replied:
(I'd be happy with big long names like '`2way_widening_outerproduct` :))

banach-space replied:
I think the key question is - are there any non-widening 2way or 4 way Ops?

c-rhodes replied:
> I think the key question is - are there any non-widening 2way or 4 way Ops?

No there aren't.

banach-space commented on mlir/include/mlir/Dialect/ArmSME/Transforms/Passes.td:
```diff
@@ -122,4 +122,43 @@ def TileAllocation
   let dependentDialects = ["func::FuncDialect"];
 }
 
+def OuterProductWidening
```
Hm, the pass doesn't "widen" the outer products, it "fuses"/"combines" them into one. WDYT?

c-rhodes replied:
see what you mean, suppose could rename to `OuterProductFusion`?

banach-space replied:
Yeah, or `SMEOuterProductFusion`.

MacDue replied:
+1 to any name with fuse or fusion (sounds cool, and more descriptive :smile:)

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+//
+// Becomes:
+//
+//  %a_packed = arm_sve.zip %a0, %a1 : vector<[8]xf16> to vector<[8]xf16>
```
zip1?

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+    arm_sme::OuterProductOp op1 = acc.getDefiningOp<arm_sme::OuterProductOp>();
+    arm_sme::OuterProductOp op2 = op;
+    if (!op1)
+      return rewriter.notifyMatchFailure(op,
+                                         "defining op of accumulator operand "
+                                         "must be an 'arm_sme.outerproduct'");
```
Could we add test for this and other cases where there's a failure? IIRC, these diagnostics are not printed unless using `--debug`?

c-rhodes replied:
that's right these are printed with debug, I think I've looked at using these diagnostics in tests before but it wasn't viable. At the very least I can add the tests, I was intending to and forgot.

banach-space replied:
yeah, tests with `--debug` would be horrible 😅

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+      return rewriter.notifyMatchFailure(
+          op, "combining kind (add or sub) of outer products must match");
+
+    if (!llvm::hasSingleElement(op1->getUses())) {
```
[nit] `hasOneUse`?

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+          op, "combining kind (add or sub) of outer products must match");
+
+    if (!llvm::hasSingleElement(op1->getUses())) {
+      // We could still widen, but if the first outer product has an
+      // accumulator it will be used as the root for tile allocation and since
+      // the widening outer product uses the same accumulator it will get
+      // assigned the same tile ID, resulting in 3 outer products and incorrect
+      // results. No accumulator would be ok, but it's simpler to prevent this
+      // altogether, since it has no benefit.
```
I think I know what you mean, but an example might simplify this. Something like (the types don't really matter for this discussion):
```mlir
// BEFORE
%a0_ext = arith.extf %a0
%b0_ext = arith.extf %b0
%0 = arm_sme.outerproduct %a0_ext, %b0_ext

%a1_ext = arith.extf %a1
%b1_ext = arith.extf %b1
%1 = arm_sme.outerproduct %a1_ext, %b1_ext, %0

%a2_ext = arith.extf %a2
%b2_ext = arith.extf %b2
%2 = arm_sme.outerproduct %a2_ext, %b2_ext, %0

// AFTER
%a_zip = arith.extf %a0, %a1
%b_zip = arith.extf %b0, %b1
%1 = arm_sme.outerproduct_2way %a_zip, %b_zip

%a2_ext = arith.extf %a2
%b2_ext = arith.extf %b2
%2 = arm_sme.outerproduct %a2_ext, %b2_ext, %1
```
Now, what exactly would go wrong in this case? :) 

Also: 
```suggestion
      // We could still widen, but since the first outer product has an
      // accumulator it will be used as the root for tile allocation and since
      // the widening outer product uses the same accumulator it will get
      // assigned the same tile ID, resulting in 3 outer products and incorrect
      // results. No accumulator would be ok, but it's simpler to prevent this
      // altogether, since it has no benefit.
```

The current wording reads to me as "if it happens that the first outer product has an acc". Where's it's more like "the first OP has an accumulator, hence". #english-not-my-first-lang :)

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+              "no benefit to widening");
+    }
+
+    auto nxnxv4i32 =
+        VectorType::get({4, 4}, rewriter.getI32Type(), {true, true});
+    auto nxnxv4f32 =
+        VectorType::get({4, 4}, rewriter.getF32Type(), {true, true});
+    auto nxv4i16 = VectorType::get({4}, rewriter.getI16Type(), true);
+    auto nxv4f16 = VectorType::get({4}, rewriter.getF16Type(), true);
+    auto nxv4bf16 = VectorType::get({4}, rewriter.getBF16Type(), true);
+    if ((failed(
+             isWidenable<arith::ExtFOp>(rewriter, op1, nxnxv4f32, nxv4f16)) ||
+         failed(
+             isWidenable<arith::ExtFOp>(rewriter, op2, nxnxv4f32, nxv4f16))) &&
+        (failed(
+             isWidenable<arith::ExtFOp>(rewriter, op1, nxnxv4f32, nxv4bf16)) ||
+         failed(
+             isWidenable<arith::ExtFOp>(rewriter, op2, nxnxv4f32, nxv4bf16))) &&
+        (failed(
+             isWidenable<arith::ExtSIOp>(rewriter, op1, nxnxv4i32, nxv4i16)) ||
+         failed(
+             isWidenable<arith::ExtSIOp>(rewriter, op2, nxnxv4i32, nxv4i16))) &&
+        (failed(
+             isWidenable<arith::ExtUIOp>(rewriter, op1, nxnxv4i32, nxv4i16)) ||
+         failed(
+             isWidenable<arith::ExtUIOp>(rewriter, op2, nxnxv4i32, nxv4i16))))
+      return failure();
```
The types that are defined at the top of this block (e.g. `nxnxv4i32`) are only used within the block itself. Perhaps all of this should be a function? Also, it's not clear why checking only these selected types. So, moving this to a dedicated function and adding some comments would be welcome :)

banach-space commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+
+private:
+  template <typename ExtOp>
+  LogicalResult isWidenable(PatternRewriter &rewriter,
```
Could you add a comment defining "widenable"? I think that this confusing as it's not about "widening" but more about "fusing". And this is trying to identify ops suitable for fusion? Also, what's the relation between `resultType` and `inputType`? Are these from a single Outer Product?

c-rhodes committed:
replace arm_sve.intr.zip1 with target-agnostic interleave2 intrinsic

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+                                                  insertRHS);
+    };
+
+    auto extOp = op.getLhs().getDefiningOp();
+    VectorType extSourceVectorType =
+        cast<VectorType>(extOp->getOperand(0).getType());
+    VectorType widenedVectorType =
+        VectorType::Builder(extSourceVectorType)
+            .setDim(0, extSourceVectorType.getShape()[0] * 2);
+    auto lhs = packInputs(widenedVectorType,
+                          op1.getLhs().getDefiningOp()->getOperand(0),
+                          op2.getLhs().getDefiningOp()->getOperand(0));
+    auto rhs = packInputs(widenedVectorType,
+                          op1.getRhs().getDefiningOp()->getOperand(0),
+                          op2.getRhs().getDefiningOp()->getOperand(0));
```
May want to check the `getDefiningOp()` results, it can return null (e.g. for function arguments, loop iter_args, etc).

MacDue commented on mlir/lib/Dialect/ArmSME/Transforms/OuterProductWidening.cpp:
```diff
@@ -0,0 +1,238 @@
+
+    Value lhsMask, rhsMask;
+    if (op1.getLhsMask() || op2.getLhsMask()) {
+      if (!(op1.getLhsMask() && op2.getLhsMask()))
+        return rewriter.notifyMatchFailure(
+            op, "unsupported masking, either both outerproducts are masked "
+                "or neither");
```
Change the condition to:
```c++
 if (bool(op1.getLhsMask()) == bool(op2.getLhsMask()))
```
and lift this outside the (`if (op1.getLhsMask() || op2.getLhsMask()) {`), I think that'd make this a little clearer (and avoid a level of nesting).

c-rhodes commented:
Rebased and updated to use target-agnostic interleave2 intrinsic instead of zip1.

c-rhodes commented on mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td:
```diff
@@ -814,6 +814,649 @@ let arguments = (ins
+  ArmSME_Op<mnemonic, [
+    ArmSMETileOpInterface,
+    AttrSizedOperandSegments,
+    AllTypesMatch<["lhs", "rhs"]>,
```
the last trait in the list verifies that.

c-rhodes committed:
Address comments. Changes:

- hasOneUse
- Negative tests for pass
- move types and isWidening checks to isSupported method
- negative test for unsupported type
- move masking check
- op1.erase() -> rewriter.eraseOp(op1);
- braces round dangling-else
- TypeSwitch
- rename isWidenable and add comments for clarity
- add TODO/REDEFINE for QEMU bug to make it clearer
- s/widening/fusion/g
- drop wide from op names

banach-space reviewed (commented):
Few more comments.

Overall I'm happy with the overall design and the execution. I am away next week -  don't wait for me to +1 this. As long as my comments and comments from @MacDue are addressed, I am happy for this to be merged.

c-rhodes commented on mlir/test/Integration/Dialect/Vector/CPU/ArmSME/test-outerproduct-f16f16f32.mlir:
```diff
@@ -0,0 +1,100 @@
+  %0 = vector.outerproduct %a0_ext, %b0_ext, %acc : vector<[4]xf32>, vector<[4]xf32>
+  %1 = vector.outerproduct %a1_ext, %b1_ext, %0 : vector<[4]xf32>, vector<[4]xf32>
+
+  // CHECK:      (  79,  95, 111, 127
```
Briefly tried this and got incorrect results, was all zeroes, not sure why. But it was correct if I explicitly passed `sme128=on` to QEMU.

c-rhodes replied:
ok the problem was I called `setArmSVLBits` in the streaming function which is undefined behaviour. Must be called before calling into func that uses vector ops (thanks @MacDue)

c-rhodes committed:
use setArmSVLBits in integration test to set SVL to 128

MacDue reviewed (approved):
LGTM! I have a few final nits, but nothing major :)

c-rhodes committed:
Address comments. Changes:

- Backticks in constraint message.
- Fix comment.
- Remove ArmSVE from libs now zip1 is gone.
- rename 'widenedType' to 'inputTypeX2' and move to 'packInputs'.
- Use getIn for arith ext ops.
- Add test to verify chain of 4 outer products are fused into 2 2-way
  outer products.

c-rhodes commented:
> LGTM! I have a few final nits, but nothing major :)

Believe I've addressed all comments now, thanks for reviewing! Andrzej is off this week so I'll land this tomorrow unless there's any further comments by then. Cheers.
